proto/remote_execution.proto

```diff
@@ -2199,6 +2199,27 @@ message UsageStats {
   // Total number of CPU-nanoseconds consumed by the task.
   int64 cpu_nanos = 2;
 
+  // CPU usage samples. This can be useful to report summaries such as "the task
+  // peaked at 3000mcpu, and was at peak for at least 10% of its execution
+  // duration."
+  //
+  // Buckets represent total cpu-millis used (e.g. 0-100 cpu-millis), and counts
+  // represent the sum of the sampling interval durations (in seconds) where the
+  // average milliCPU usage fell in that range.
+  //
+  // For example, let's say the executor samples CPU usage every 100ms, and it
+  // executed a task that ran for 200ms, and during the first 100ms we observed
+  // 3140 total cpu-millis used, and during the second 200ms we observed 111
+  // total cpu-millis used. Then a valid histogram representing this execution
+  // could be:
+  //
+  // { buckets: [100, 3100], counts: [0.1, 0.1], bucket_width: 100 }
+  //
+  // Note: this histogram does not convey any information about *when* the CPU
+  // usage occurred; it only reports information on how frequently we observed
+  // each range of CPU usages.
```

Comment:
Another option that preserves timing information would be to report CPU usage at every 100ms interval and then if a task runs for a very long time we can "collapse" intervals so that we only report every second or every minute, to keep the proto reasonably small. I figured this approach would be slightly more complex but it may be more useful / interesting to report it this way? curious if anyone has thoughts here

Reply:
delta compression would probably work well here if your CPU stats are quantized and the period is fixed

Reply:
You can also use run length encoding for the intervals if you want.

Anyway before commiting to a format, you probably want to run a long-ish task, gather the data and encode it and see if your format makes sense.

Reply:
For users, some timeline information is definitely useful (compared to a histogram), but I suspect that when they really need this data (like a large test), they'll want to know which process or function used CPU when. That means running some sort of profiler. So I think less precise data here is more than enough to give them an overview. 100ms windows are probably too small. 1s seems good. Maybe 10s is all you need.

I like the idea of delta compression, though it does make the data a bit harder to use. If the numbers can be reasonably small, each sample could be 2-3 bytes. Though, you could get also small numbers if samples were rounded to the nearest deci-period. For example, let's say the period is 1s. Instead of a sample value of 3517ms indicating that the action used 3.517 cores during that period, use a sample of 35 deciseconds to indicate that the action used ~3.5 cores during that period.

I'm just guessing, but RLE is probably not worth it. Analyzing the data in something like a SQL query becomes way harder then. Also, I think it requires a repeated message instead of a primitive field, which means you don't benefit from packed fields.

Reply:
For some more background: the main motivation for this PR is to start collecting more data in order to improve our automatic task sizing. This is sort of a pre-requisite for being able to apply hard-limits on task resource usage, since we don't want to apply automatic hard limits if we aren't confident about what the limits should be. There is more discussion in this issue: https://github.com/buildbuddy-io/buildbuddy-internal/issues/3965

We currently estimate task sizes by using average CPU (total cpu-milliseconds used divided by execution duration), which turns out to be a poor estimate for certain types of workloads - e.g. some VM workloads spend a lot of work up-front decompressing docker images using multiple cores, and then once the image is decompressed, the test is single-threaded and/or IO bound, resulting in a lot less CPU being used throughout the test. So using the average CPU to auto-size these workloads doesn't work very well.

One idea we discussed was recording the CPU usage over time and then somehow improving our task sizing model using that information. Our current idea is to use p90 CPU usage (this is just a starting point which will likely have some drawbacks; maybe some more analysis could be done to understand how to more effectively use this info).

Encoding aside, I'm curious if you have any thoughts about this overall approach?

But in terms of the implementation here I do agree that we could probably just collect samples in 1s intervals and report every single sample as a plain old int32. That's just 14 KB per timeseries per hour of execution time.

Reply:
I meant to talk with you about this at Bazelcon, but then I forgot. I'll set up a VC for tomorrow. I have a bunch of questions before I can have any well informed thoughts on this.

Reply:
We chatted a bit offline and brainstormed various ideas about how to improve the CPU task sizing situation, but still think that measuring this CPU timeline is useful.